refactor(core): Throw an error when the document is not initialized.

JeanMeche opened the pull request:
In case the document is accessed but not available we should throw ASAP an error to prevent non explicit errors.

See #50138, where this issue happens. 


## PR Type
What kind of change does this PR introduce?

- [x] Bugfix

alan-agius4 commented on packages/core/src/render3/interfaces/document.ts:
```diff
@@ -47,6 +49,11 @@ export function getDocument(): Document {
     return document;
   }
+
+  if(typeof ngDevMode === 'undefined' || ngDevMode && !getDocument()) {
```
I think this is case we should always throw irrespective if we are in dev mode or not.

JeanMeche replied:
Yeah, definitely.

JeanMeche commented:
> This also does not solve the true problem described in #50138

Yes, I understand it is not a true fix, just a mere helper to look in the right direction<

alan-agius4 commented on packages/core/src/render3/interfaces/document.ts:
```diff
@@ -47,6 +49,14 @@ export function getDocument(): Document {
   } else if (typeof document !== 'undefined') {
     return document;
   }
+
```
The logic from this points onwards can be deleted in favor of the throw. If is redundant to call `getDocument`.

alan-agius4 commented:
> > This also does not solve the true problem described in #50138
> 
> Yes, I understand it is not a true fix, just a mere helper to look in the right direction.

I suggest that in this case, the commit message `fixes` is removed.

JeanMeche committed:
refactor(core): Throw an error when the document is not initialized.

In case the document is accessed but not available we should throw ASAP an error to prevent non explicit errors.

alan-agius4 commented on packages/core/src/render3/interfaces/document.ts:
```diff
@@ -47,10 +49,15 @@ export function getDocument(): Document {
+      (typeof ngDevMode === 'undefined' || ngDevMode) &&
+          `The document object is not available in this context. Make sure the DOCUMENT injection token is provided.`);
+
   // No "document" can be found. This should only happen if we are running ivy outside Angular and
```
Nit: maybe move the comment before the throw.

alan-agius4 reviewed (approved):
LGTM

AndrewKushnir reviewed (approved):
@JeanMeche looks great, thanks! 👍 

Reviewed-for: fw-core, public-api

dylhunn reviewed (approved):
reviewed-for: public-api

AndrewKushnir commented:
[TGP](https://test.corp.google.com/OCL:529596253:BASE:529596204:1683258983931:752d192d).

AndrewKushnir commented:
**Caretaker note**: TGP is "green".

alxhub commented:
This PR was merged into the repository by commit 02147394880fe7b950d700f332758f5c09cb5c27.
